### PR TITLE
hv: some improvements to msi.c and msix.c

### PR DESCRIPTION
**Pull request by ZideChen0**
- Upon setting or clearing MSI-X Enable bit, it does MSI-X remap for all
  entries in the MSI-X table. One of the negative impacts of doing this
  is that it could create unused ptdev remapping entries. This patch
  changes the behavior to only do remap for those entries that are not
  masked from Vector Control.

- When disabling MSI-X or masking MSI-X entries, no need to call
  ptdev_msix_remap() to do remapping. This patch simplifies the remapping
  logic by replacing the existing remapping functions with vmsix_enable(),
  vmsix_disable() and vmsix_update_entry(). Similar changes to msi.c

- The guest could write the MSI-X Table fields in DWORD or QWORD. For
  example the Message Address could be a QWORD write or two DWORD writes,
  the Message Data could be a DWORD write or part of a QWORD write that
  is combined with the Vector Control. This patch captures all the
  possible combinations when the guest is altering the MSI-X Table.

- Misc. fixes: correctly initialize the emulated MSI-X Table, etc.

Tracked-On: #1568
Signed-off-by: Zide Chen <zide.chen@intel.com>
Reviewed-by: Li, Fei1 <fei1.li@intel.com>

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
1845<br>2 new potential violations to the coding guideline detected. Please help check. Thanks!  <br><br>1.filename:/hypervisor/dm/vpci/msix.c&nbsp;&nbsp;function:vmsix_disable_entry&nbsp;&nbsp;offset:53:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Cast from integral type to pointer. : (unsigned long to struct msix_table_entry*): ( struct msix_table_entry * ) hva<br><br>2.filename:/hypervisor/dm/vpci/msix.c&nbsp;&nbsp;function:vmsix_update_entry&nbsp;&nbsp;offset:75:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Cast from integral type to pointer. : (unsigned long to struct msix_table_entry*): ( struct msix_table_entry * ) hva<br><br>

File: hypervisor/dm/vpci/msi.c [hunk 1]
```diff
@@ -89,15 +79,39 @@ static int vmsi_remap(struct pci_vdev *vdev, bool enable)
...
+}
+
+static int vmsi_enable(struct pci_vdev *vdev)
+{
+	union pci_bdf pbdf = vdev->pdev.bdf;
+	uint32_t msgoff = vdev->msi.capoff + PCIR_MSI_CTRL;
+	uint32_t msgctrl;
+	int ret = 0;
+
+	ret = vmsi_update(vdev);
```
Comment (anthonyzxu) [inline hypervisor/dm/vpci/msi.c]: do we still need to call vmsi_update in vmsi_enable?

File: hypervisor/dm/vpci/msi.c [hunk 2]
```diff
@@ -132,11 +145,14 @@ static int vmsi_cfgwrite(struct pci_vdev *vdev, uint32_t offset, uint32_t bytes,
...
-			(void)vmsi_remap(vdev, enable);
+			if ((val & PCIM_MSICTRL_MSI_ENABLE) != 0U) {
+				(void)vmsi_enable(vdev);
+			} else {
+				vmsi_disable(vdev);
+			}
 		} else {
 			if (message_changed && ((msgctrl & PCIM_MSICTRL_MSI_ENABLE) != 0U)) {
-				(void)vmsi_remap(vdev, true);
+				(void)vmsi_update(vdev);
```
Comment (anthonyzxu) [inline hypervisor/dm/vpci/msi.c]: would this lead to physical MSI change w/ MSI enabled?

File: hypervisor/dm/vpci/msi.c (hunk 1 above)
Comment (ZideChen0) [reply hypervisor/dm/vpci/msi.c]: Yes, currently vmsi_update() is called in 2 cases:
1) When MSI Enable bit is being set
2) Message Data/Addr is being changed when MSI is enabled

As Fei suggested, we can choose to remove 1) and in 2), we need to call vmsi_update() regardless MSI Enable is set or not. Similar logic applies to msix.c as well.

Comment (anthonyzxu) [reply hypervisor/dm/vpci/msi.c]: Is it safe to call vmsi_update with MSI enable?
if it is not safe, even guest does it, HV should avoid it.

**Pull request closed without merging**